repair: Reduce hints and batchlog flush

asias opened the pull request:
The hints and batchlog flush requests are issued to all nodes for each repair request when tombstone_gc repair mode is used.

The amount of such flush requests is high when all nodes in the cluster run repair. It is observed it takes a long time, up to 15s, for a repair request to finish such a flush request.

To reduce overhead of the flush, each node caches the flush and only executes the real flush when some time has passed. It is safe to do so before the real flush_time is returned. Repair uses the smallest flush_time from peers as the repair time.

The nice thing about the cache on the receiver side is that all senders can hit the cache. It is better than cache on the sender side.

A slightly smaller flush_time compared to the real flush time will be used with the benefits of significantly dropped hints and batchlog flush. The tradeoff is reasonable.

Fixes #20259

Performance improvement. No backports.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11065/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11065/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3168/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/120/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11191/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11065/)
#### Build Details:
- Duration: 2 hr 56 min
- Builder: spider4.cloudius-systems.com

vladzcloudius commented on repair/row_level.cc:
```diff
@@ -2290,32 +2290,45 @@ future<repair_update_system_table_response> repair_service::repair_update_system
-        rlogger.warn("repair[{}]: Failed to process repair_flush_hints_batchlog_request from node={}, target_hosts={}, {}",
-                req.repair_uuid, from, req.target_nodes, std::current_exception());
-        throw;
+    auto constexpr max_repair_hints_batchlog_flush_time = std::chrono::seconds(60);
```
Please, make it configurable and LiveUpdate

asias replied:
It is already a config in my local version. I did not push it yet.

vladzcloudius reviewed (changes requested):
Please, make sure to allow configuring corresponding flush frequency

vladzcloudius commented:
@asias could you, please, back this PR up with testing numbers: repair time before and after.
This is a performance optimization - let's see what it gives us.

asias commented:
Changes in 089751163767cc8b8d512873f3c85f3d0d634290:
- Added get_last_replay logic to avoid flush in case batchlog manager has replayed
- Integrated Botond's BM change.
- Added tests test_repair.py 
- Rebased to latest master

asias commented:
> @asias could you, please, back this PR up with testing numbers: repair time before and after. This is a performance optimization - let's see what it gives us.

I saw 30X improvement in my test. 

    Tests: 2 nodes, with 1s batchlog delay:
    
    Before:
       Repair nr_repairs=20 cache_time_in_ms=0 total_repair_duration=40.04245328903198
    
    After:
       Repair nr_repairs=20 cache_time_in_ms=5000 total_repair_duration=1.252073049545288

denesb commented on repair/repair.cc:
```diff
@@ -397,30 +399,43 @@ static future<bool> flush_hints(repair_service& rs, repair_uniq_id id, replica::
-                    (void)resp; // nothing to do with response yet
+                    if (resp.flush_time == gc_clock::time_point()) {
+                        // This means the node does not support sending flush_time back. Use the time when the flush is requested for flush_time.
+                        rlogger.warn("repair[{}]: Got empty flush_time from node={}. Please upgrade the node={}.", uuid, node, node);
```
This is unnecessary and will spam the logs during upgrade. The operator will upgrade all nodes eventually anyway.

asias replied:
We can make it debug to make it less noisy.

denesb commented on db/config.cc:
```diff
@@ -972,6 +972,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
             " This can reduce the amount of data repair has to process.")
     , repair_partition_count_estimation_ratio(this, "repair_partition_count_estimation_ratio", liveness::LiveUpdate, value_status::Used, 0.1,
         "Specify the fraction of partitions written by repair out of the total partitions. The value is currently only used for bloom filter estimation. Value is between 0 and 1.")
+    , repair_hints_batchlog_flush_cache_time_in_ms(this, "repair_hints_batchlog_flush_cache_time_in_ms", value_status::Used, 60 * 1000, "The repair hints and batchlog flush request cache time. Setting 0 disables the flush cache.")
```
The description doesn't explain what this flush request cache is.

denesb commented on repair/row_level.cc:
```diff
@@ -2277,36 +2277,86 @@ future<repair_update_system_table_response> repair_service::repair_update_system
+        });
+        updated = true;
+    } else {
+        utils::get_local_injector().set_parameter("repair_flush_hints_batchlog_handler", "skip_flush", fmt::to_string(flush_time));
```
This is a large change. Can't we eliminate some of the churn by just flipping the if and having an early return?

asias replied:
If you look at the patch with --ignore-all-space, the diff will be much smaller.  Early return means we will print the finish log twice.

denesb replied:
Yes, I did that and it helped. Still, I usually prefer not having to do that.

denesb commented:
This PR includes code I wrote, so I would like at least one other review.

asias commented:
Changes in v2: 
- Updated the help message for the option
- Change one log from warning to debug

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11483/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11483/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: topology_custom/test_repair
#### Failed Tests (8/151):
- [test_batchlog_flush_in_repair_with_cache.debug.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.10)
- [test_batchlog_flush_in_repair_with_cache.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.2)
- [test_batchlog_flush_in_repair_with_cache.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.3)
- [test_batchlog_flush_in_repair_with_cache.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.4)
- [test_batchlog_flush_in_repair_with_cache.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.5)
- [test_batchlog_flush_in_repair_with_cache.debug.7](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.7)
- [test_batchlog_flush_in_repair_with_cache.debug.8](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.8)
- [test_batchlog_flush_in_repair_with_cache.debug.9](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11483/testReport/junit/%28root%29/test_repair/test_batchlog_flush_in_repair_with_cache_debug_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_batchlog_flush_in_repair_with_cache.debug.9)
#### Build Details:
- Duration: 1 hr 5 min
- Builder: spider3.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: topology_custom/test_repair
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/483/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11554/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3535/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
#### Failed Tests (24/36770):
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
#### Build Details:
- Duration: 3 hr 58 min
- Builder: spider7.cloudius-systems.com

asias commented on test/topology_custom/test_repair.py:
```diff
@@ -167,3 +168,63 @@ async def test_repair_succeeds_with_unitialized_bm(manager):
+    """
+    nr_repairs_per_node = 3
+    nr_repairs = 2 * nr_repairs_per_node
+    total_repair_duration = 0;
```
done

asias commented on test/topology_custom/test_repair.py:
```diff
@@ -167,3 +168,63 @@ async def test_repair_succeeds_with_unitialized_bm(manager):
+    total_repair_duration += duration
+
+    assert (int(params['issue_flush']) > 0)
+    if (cache_time_in_ms > 0):
```
yes, done

asias commented on repair/repair.cc:
```diff
@@ -397,30 +399,43 @@ static future<bool> flush_hints(repair_service& rs, repair_uniq_id id, replica::
                     throw;
                 }
             });
+            if (!times.empty()) {
+                auto it = std::min_element(times.begin(), times.end());
+                flush_time = *it;
+            }
```
The participants will include the node itself at least, so it will not be empty. Also, it will be safer to set the flush time to zero by default.  If no real flush time is returned, we should not move the flush time to now.

asias commented on repair/row_level.cc:
```diff
@@ -2277,36 +2277,86 @@ future<repair_update_system_table_response> repair_service::repair_update_system
+                    if (issue_flush) {
+                        co_await _bm.local().do_batch_log_replay(db::batchlog_manager::post_replay_cleanup::no);
+                        utils::get_local_injector().set_parameter("repair_flush_hints_batchlog_handler", "issue_flush", fmt::to_string(flush_time));
+                    }
```
I added the issue_flush to the log to let us know if it the flush is issued or not. It is useful to know the start and the end of the flush operation. It helped us to figure out the long flush issue in the past.

asias commented:
Changes in 838b1ef4ff20a9cb3bf0f0c497a4e67f5a0f7ef9:
- Nit fix in test
- Updated the log to show if the flush is issued or not

avikivity commented on db/config.cc:
```diff
@@ -850,6 +850,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , batchlog_replay_throttle_in_kb(this, "batchlog_replay_throttle_in_kb", value_status::Unused, 1024,
         "Total maximum throttle. Throttling is reduced proportionally to the number of nodes in the cluster.")
+    , batchlog_replay_cleanup_after_replays(this, "batchlog_replay_cleanup_after_replays", value_status::Unused, 100,
+        "Clean up batchlog memtable after every N replays. Replays are issued on a timer, every ring_delay_ms milliseconds. So if batchlog_replay_cleanup_after_replays is set to 100, the batchlog memtable is flushed every 100 * ring_delay_ms milliseconds")
```
ring_delay_ms is very low in tests (0) and is mostly meaningless. I think we should redefine the timer, and set this to also be a duration.

avikivity replied:
Make it LiveUpdate too.

asias replied:
The replay is issued every 60 seconds except the first replay during boot. Only the first replay sleeps ring_delay_ms. @denesb Can you double check and confirm?

```
satic constexpr uint32_t replay_interval = 60 * 1000; // milliseconds

   auto delay = _delay;
    while (!_stop.abort_requested()) {
        try {
            co_await sleep_abortable(delay, _stop);
        } catch (sleep_aborted&) {
            co_return;
        }
        try {
            auto cleanup = post_replay_cleanup::no;
            if (++replay_counter >= _replay_cleanup_after_replays) {
                replay_counter = 0;
                cleanup = post_replay_cleanup::yes;
            }
            co_await do_batch_log_replay(cleanup);
        } catch (seastar::broken_semaphore&) {
            if (_stop.abort_requested()) {
                co_return;
            }
            on_internal_error_noexcept(blogger, fmt::format("Unexcepted exception in batchlog reply: {}", std::current_exception()));
        } catch (...) {
            blogger.error("Exception in batch replay: {}", std::current_exception());
        }
        delay = std::chrono::milliseconds(replay_interval);
    } 

```

So I think it makes sense to set  batchlog_replay_cleanup_after_replays to be a number of replay. We can make replay_interval configurable as well in a follow up if it is needed.

avikivity commented on db/batchlog_manager.hh:
```diff
@@ -69,6 +69,8 @@ private:
     future<> _loop_done;
 
+    gc_clock::time_point _last_replay;
+
```
Why gc_clock? It's not related to gc. Better to use lowres_clock.

asias replied:
It is related to gc because we use the last replay time to decide the repair_time which is used to do gc.

avikivity commented:
Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).

asias commented:
Changes in d2e1b84975b903d08e355b966400b91e76cc64dc:
- Modified help message for batchlog_replay_cleanup_after_replays
- Made batchlog_replay_cleanup_after_replays liveupdate
- Rebased

asias commented:
> Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).

We need to extend existing SCT tests to have batchlog writes in the workload.  It seems we have no batchlog related tests in SCT.  @mykaul Could you help to assign?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11571/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11571/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: topology_custom/test_repair
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11571/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3570/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11587/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11571/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/518/)
#### Failed Tests (24/36885):
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy])
- [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11571/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy])
#### Build Details:
- Duration: 3 hr 41 min
- Builder: spider7.cloudius-systems.com

asias commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/batchlog_manager_test 🔹 topology_custom/test_repair ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/483/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11554/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3535/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11510/)
> 
> #### Failed Tests (24/36770):
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-SizeTieredCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-TimeWindowCompactionStrategy%5D)
> * [test_compaction_delete_tombstone_gc[repair-LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11510/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_compaction_delete_tombstone_gc_repair_LeveledCompactionStrategy__2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_compaction_delete_tombstone_gc%5Brepair-LeveledCompactionStrategy%5D)
> 
> #### Build Details:
> * Duration: 3 hr 58 min
> * Builder: spider7.cloudius-systems.com

We need to adjust the dtest. PR is sent https://github.com/scylladb/scylla-dtest/pull/4934.

mykaul commented:
> > Looks generally good. I'd like to see a realistic measurement of the improvement (not on a laptop cluster).
> 
> We need to extend existing SCT tests to have batchlog writes in the workload. It seems we have no batchlog related tests in SCT. @mykaul Could you help to assign?

@pehala ?
(I assume it's indirectly part of https://github.com/scylladb/scylla-cluster-tests/issues/8446 )

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11693/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11693/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: topology_custom/test_repair
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11693/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/623/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3678/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11692/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11693/)
#### Build Details:
- Duration: 4 hr 13 min
- Builder: spider8.cloudius-systems.com

pehala commented:
> @pehala ?
> (I assume it's indirectly part of [scylladb/scylla-cluster-tests#8446](https://github.com/scylladb/scylla-cluster-tests/issues/8446) )

Yup, I will add it as as a note to the issue

asias commented:
Now the CI has passed and we have a plan for the sct test.  Can we get this merged? @denesb

denesb commented:
> Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb

There is an unaddressed comment from @avikivity : https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161.

asias commented:
> > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> 
> There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).

As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test.  If the batchlog sct perf test finds any issue later we can fix it as a follow up.

mykaul commented:
> > > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> > 
> > 
> > There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).
> 
> As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test. If the batchlog sct perf test finds any issue later we can fix it as a follow up.

As long as we are careful not to merge its backports (currently we do not have the labels for it, but we could), then I think we are fine with merging this.

denesb commented:
> > > Now the CI has passed and we have a plan for the sct test. Can we get this merged? @denesb
> > 
> > 
> > There is an unaddressed comment from @avikivity : [#20260 (comment)](https://github.com/scylladb/scylladb/pull/20260#issuecomment-2346328161).
> 
> As I mentioned above, the SCT with batchlog and repair is being worked on which will take a while. Yes, it is nice to see the big cluster results. We already have small scale tests showing significant improvement. Are we going to block the merge and wait for the SCT results? Isn't better to merge it now so it could gain more tests in our testing pipeline while waiting for the batchlog sct test. If the batchlog sct perf test finds any issue later we can fix it as a follow up.

It is possible that a change which is intended as and advertised as a performance improvement turns out to be a performance regression later. In that case, merging it has no value. That is why we ask for measurements up-front and not later.

In any case @avikivity requested the test, so I'll let him decide if he is fine with the tests being done post-merge.

asias commented:
There is no update from QA on the test for a while, so I decided to hack and do the SCT test myself.

Config: 3 i3en.xlarge nodes, 10 keyspaces each has 1 table, 1800 repair requests in total.

Here is the result:

1) Repair Before: 9754s
    
f:nemesis.py l:1758 c:sdcm.nemesis p:INFO  > HJ: Finished repair on db nodes in parallel time_elapsed=9754s nr_repairs=1800 use_mgmt=False repair_ratio=2 keyspaces=['ks0', 'ks1', 'ks2', 'ks3', 'ks4', 'ks5', 'ks6', 'ks7', 'ks8', 'ks9'] 
    
2) Repair After: 752s 
    
f:nemesis.py l:1758 c:sdcm.nemesis p:INFO  > HJ: Finished repair on db nodes in parallel time_elapsed=752s nr_repairs=1800 use_mgmt=False repair_ratio=2 keyspaces=['ks0', 'ks1', 'ks2', 'ks3', 'ks4', 'ks5', 'ks6', 'ks7', 'ks8', 'ks9']

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12857/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12857/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/batchlog_manager_test
:small_blue_diamond: topology_custom/test_repair
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/17/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4515/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12541/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1445/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12857/)
#### Build Details:
- Duration: 4 hr 29 min
- Builder: spider5.cloudius-systems.com
